File: tests/bin/run-tests.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Simple test runner for Ruby backend. Mirrors run-tests.py.

require 'set'

# Find all .tests files recursively
def find_test_files(directory)
  result = []
  Dir.glob(File.join(directory, '**', '*.tests')).each do |f|
    result << f
  end
  result.sort
end

# Parse a .tests file. Returns array of [name, input, expected, line_num]
def parse_test_file(filepath)
  tests = []
  lines = File.read(filepath).split("\n")
  i = 0
  n = lines.length
  while i < n
    line = lines[i]
    # Skip comments and blank lines
    if line.start_with?('#') || line.strip.empty?
      i += 1
      next
    end
    # Test header
    if line.start_with?('=== ')
      name = line[4..].strip
      start_line = i + 1
      i += 1
      # Collect input until ---
      input_lines = []
      while i < n && lines[i] != '---'
        input_lines << lines[i]
        i += 1
      end
      # Skip ---
      i += 1 if i < n && lines[i] == '---'
      # Collect expected until ---, next test, or EOF
      expected_lines = []
      while i < n && lines[i] != '---' && !lines[i].start_with?('=== ')
        expected_lines << lines[i]
        i += 1
      end
      # Skip --- end marker
      i += 1 if i < n && lines[i] == '---'
      # Strip trailing blank lines from expected
      expected_lines.pop while !expected_lines.empty? && expected_lines[-1].strip.empty?
      test_input = input_lines.join("\n")
      test_expected = expected_lines.join("\n")
      tests << [name, test_input, test_expected, start_line]
    else
      i += 1
    end
  end
  tests
end

# Normalize whitespace for comparison
def normalize(s)
  s.encode('UTF-8', invalid: :replace, undef: :replace, replace: '?').split.join(' ')
end

# Run a single test. Returns [passed, actual, error_msg]
def run_test(test_input, test_expected)
  # Check for @extglob directive
  extglob = false
  if test_input.start_with?("# @extglob\n")
    extglob = true
    test_input = test_input[("# @extglob\n").length..]
  end

  begin
    nodes = parse(test_input, extglob)
    actual = nodes.map(&:to_sexp).join(' ')
  rescue ParseError => e
    return [true, '<error>', nil] if normalize(test_expected) == '<error>'
    return [false, '<parse error>', e.message]
  rescue StandardError => e
    return [false, '<exception>', "#{e.class}: #{e.message}"]
  end

  # If we expected an error but got a successful parse, that's a failure
  if normalize(test_expected) == '<error>'
    return [false, actual, 'Expected parse error but got successful parse']
  end

  expected_norm = normalize(test_expected)
  actual_norm = normalize(actual)
  if expected_norm == actual_norm
    [true, actual, nil]
  else
    [false, actual, nil]
  end
end

def main
  # Find test directory
  script_dir = File.dirname(File.expand_path(__FILE__))
  test_dir = File.dirname(script_dir) # tests/
  repo_root = File.dirname(test_dir)

  # Parse arguments
  verbose = false
  filter_pattern = nil
  i = 0
  while i < ARGV.length
    arg = ARGV[i]
    if arg == '-v' || arg == '--verbose'
      verbose = true
    elsif arg == '-f' || arg == '--filter'
      i += 1
      filter_pattern = ARGV[i] if i < ARGV.length
    elsif File.exist?(arg)
      test_dir = arg
    end
    i += 1
  end

  # Find and run tests
  start_time = ::Time.now
  total_passed = 0
  total_failed = 0
  failed_tests = []

  test_files = if File.file?(test_dir)
                 [test_dir]
               else
                 find_test_files(test_dir)
               end

  test_files.each do |filepath|
    tests = parse_test_file(filepath)
    rel_path = filepath.sub("#{repo_root}/", '')

    tests.each do |name, test_input, test_expected, line_num|
      # Apply filter
      if filter_pattern
        next unless name.include?(filter_pattern) || rel_path.include?(filter_pattern)
      end

      # Treat <infinite> as <error>
      effective_expected = test_expected
      effective_expected = '<error>' if normalize(test_expected) == '<infinite>'

      passed, actual, error_msg = run_test(test_input, effective_expected)

      if passed
        total_passed += 1
        puts "PASS #{rel_path}:#{line_num} #{name}" if verbose
      else
        total_failed += 1
        failed_tests << [rel_path, line_num, name, test_input, test_expected, actual, error_msg]
        puts "FAIL #{rel_path}:#{line_num} #{name}" if verbose
      end
    end
  end

  elapsed = ::Time.now - start_time

  # Print summary
  puts ''
  if total_failed > 0
    puts '=' * 60
    puts 'FAILURES'
    puts '=' * 60
    failed_tests.each do |rel_path, line_num, name, inp, expected, actual, error_msg|
      puts "\n#{rel_path}:#{line_num} #{name}"
      puts "  Input:    #{inp.inspect}"
      puts "  Expected: #{expected}"
      puts "  Actual:   #{actual}"
      puts "  Error:    #{error_msg}" if error_msg
    end
    puts ''
  end

  puts "#{total_passed} passed, #{total_failed} failed in #{'%.2f' % elapsed}s"

  exit(1) if total_failed > 0
  exit(0)
end

main

File: tests/bin/run-tests.pl
#!/usr/bin/env perl
# Simple test runner for Perl backend. No dependencies beyond parable itself.

use strict;
use warnings;
use feature 'signatures';
no warnings 'experimental::signatures';
use File::Find;
use File::Basename;
use Time::HiRes qw(time);

# Import parable (it's a plain .pl file, not a module)
require 'parable.pl';

sub find_test_files ($directory) {
    my @result;
    find(sub {
        push @result, $File::Find::name if /\.tests$/;
    }, $directory);
    return sort @result;
}

sub parse_test_file ($filepath) {
    open my $fh, '<', $filepath or die "Cannot open $filepath: $!";
    my @lines = <$fh>;
    close $fh;
    chomp @lines;

    my @tests;
    my $i = 0;
    my $n = scalar @lines;

    while ($i < $n) {
        my $line = $lines[$i];
        # Skip comments and blank lines
        if ($line =~ /^#/ || $line =~ /^\s*$/) {
            $i++;
            next;
        }
        # Test header
        if ($line =~ /^=== (.*)/) {
            my $name = $1;
            my $start_line = $i + 1;
            $i++;
            # Collect input until ---
            my @input_lines;
            while ($i < $n && $lines[$i] ne '---') {
                push @input_lines, $lines[$i];
                $i++;
            }
            # Skip ---
            $i++ if $i < $n && $lines[$i] eq '---';
            # Collect expected until ---, next test, or EOF
            my @expected_lines;
            while ($i < $n && $lines[$i] ne '---' && $lines[$i] !~ /^=== /) {
                push @expected_lines, $lines[$i];
                $i++;
            }
            # Skip --- end marker
            $i++ if $i < $n && $lines[$i] eq '---';
            # Strip trailing blank lines from expected
            pop @expected_lines while @expected_lines && $expected_lines[-1] =~ /^\s*$/;

            my $test_input = join("\n", @input_lines);
            my $test_expected = join("\n", @expected_lines);
            push @tests, [$name, $test_input, $test_expected, $start_line];
        } else {
            $i++;
        }
    }
    return @tests;
}

sub normalize ($s) {
    $s =~ s/\s+/ /g;
    $s =~ s/^\s+//;
    $s =~ s/\s+$//;
    return $s;
}

sub run_test ($test_input, $test_expected) {
    # Check for @extglob directive
    my $extglob = 0;
    if ($test_input =~ s/^# \@extglob\n//) {
        $extglob = 1;
    }

    my $actual;
    eval {
        my $nodes = main::parse($test_input, $extglob);
        $actual = join(' ', map { $_->to_sexp() } @$nodes);
    };
    if ($@) {
        my $error = $@;
        if (normalize($test_expected) eq '<error>') {
            return (1, '<error>', undef);
        }
        return (0, '<parse error>', $error);
    }

    # If we expected an error but got a successful parse, that's a failure
    if (normalize($test_expected) eq '<error>') {
        return (0, $actual, 'Expected parse error but got successful parse');
    }

    my $expected_norm = normalize($test_expected);
    my $actual_norm = normalize($actual);
    if ($expected_norm eq $actual_norm) {
        return (1, $actual, undef);
    } else {
        return (0, $actual, undef);
    }
}

sub main {
    my $script_dir = dirname(__FILE__);
    my $test_dir = dirname($script_dir);  # tests/
    my $repo_root = dirname($test_dir);

    # Parse arguments
    my $verbose = 0;
    my $filter_pattern;
    my $i = 0;
    while ($i < @ARGV) {
        my $arg = $ARGV[$i];
        if ($arg eq '-v' || $arg eq '--verbose') {
            $verbose = 1;
        } elsif ($arg eq '-f' || $arg eq '--filter') {
            $i++;
            $filter_pattern = $ARGV[$i] if $i < @ARGV;
        } elsif (-e $arg) {
            $test_dir = $arg;
        }
        $i++;
    }

    # Find and run tests
    my $start_time = time();
    my $total_passed = 0;
    my $total_failed = 0;
    my @failed_tests;

    my @test_files;
    if (-f $test_dir) {
        @test_files = ($test_dir);
    } else {
        @test_files = find_test_files($test_dir);
    }

    for my $filepath (@test_files) {
        my @tests = parse_test_file($filepath);
        my $rel_path = $filepath;
        $rel_path =~ s/^\Q$repo_root\E\/?//;

        for my $test (@tests) {
            my ($name, $test_input, $test_expected, $line_num) = @$test;

            # Apply filter
            if (defined $filter_pattern) {
                next unless $name =~ /\Q$filter_pattern\E/ || $rel_path =~ /\Q$filter_pattern\E/;
            }

            # Treat <infinite> as <error>
            my $effective_expected = $test_expected;
            if (normalize($test_expected) eq '<infinite>') {
                $effective_expected = '<error>';
            }

            my ($passed, $actual, $error_msg) = run_test($test_input, $effective_expected);

            if ($passed) {
                $total_passed++;
                print "PASS $rel_path:$line_num $name\n" if $verbose;
            } else {
                $total_failed++;
                push @failed_tests, [$rel_path, $line_num, $name, $test_input, $test_expected, $actual, $error_msg];
                print "FAIL $rel_path:$line_num $name\n" if $verbose;
            }
        }
    }

    my $elapsed = time() - $start_time;

    # Print summary
    print "\n";
    if ($total_failed > 0) {
        print "=" x 60, "\n";
        print "FAILURES\n";
        print "=" x 60, "\n";
        for my $failure (@failed_tests) {
            my ($rel_path, $line_num, $name, $inp, $expected, $actual, $error_msg) = @$failure;
            print "\n$rel_path:$line_num $name\n";
            print "  Input:    '$inp'\n";
            print "  Expected: $expected\n";
            print "  Actual:   $actual\n";
            print "  Error:    $error_msg\n" if defined $error_msg;
        }
        print "\n";
    }

    printf "%d passed, %d failed in %.2fs\n", $total_passed, $total_failed, $elapsed;

    exit($total_failed > 0 ? 1 : 0);
}

main();
